dht/dht.go

```diff
@@ -618,13 +648,28 @@ func (s *Server) nodeByID(id string) *node {
 	return nil
 }
 
-func (s *Server) handleQuery(source dHTAddr, m Msg) {
+func (s *Server) handleQuery(source DHTAddr, m Msg) {
+	var (
+		newmsg *Msg
+		skip   bool
+	)
 	node := s.getNode(source, m.SenderID())
 	node.lastGotQuery = time.Now()
 	// Don't respond.
 	if s.config.Passive {
 		return
 	}
+	// Call any hooks that may apply
+	if hook, ok := s.hooks[m.Q]; ok {
+		newmsg, skip = hook(&source, node, &m)
+		if newmsg != nil {
```

Comment:
I think this is a "big picture" decision; or, performance vs. stability.

My instinct was stability; to return a new msg, because that avoids any potential side-effects if other parts of router logic log or depend on incoming messages remaining true to their state upon arrival. Message variables in the Query handlers would get overwritten but the original value should remain the same (I think?). That's a nice guarantee to have when developing the router, especially for logging, reports, and perhaps even for the blacklist. However, it comes at the cost of allocating new messages.

If the point of `torrent` is to be highly performant and/or to minimise memory allocations / GC cycles, then editing in place makes more sense, but it may lead to side-effects or even race conditions if messages are currently, or in the future, expected to be static.

Again; your party, your call. :)